Add Direct Line Speech support (#2621)
microsoft/BotFramework-WebChat: samples/06.i.direct-line-speech/README.md

# Sample - Integrating with Direct Line Speech channel

## Description

A simple page with Web Chat integrated with speech-to-text and text-to-speech feature from Direct Line Speech channel.

Direct Line Speech is a channel for communicating with a bot via a low-latency speech-focused protocol. The channel is designed for real-time speech purpose.

# Test out the hosted sample

-  [Try out MockBot](https://microsoft.github.io/BotFramework-WebChat/06.i.direct-line-speech)

# How to run

-  Fork this repository
-  Navigate to `/Your-Local-WebChat/samples/06.c.cognitive-services-speech-services-js` in command line
-  Run `npx serve`
-  Browse to [http://localhost:5000/](http://localhost:5000/)

# Things to try out

-  Click on microphone button
-  Say "123"
-  It should recognize as "123."

> -  It should speak out the next activities sent from the bot

Author: No, we don't have any DLSpeech commands, unless you count "What's the weather?"

# Code

> Jump to [completed code](#completed-code) to see the end-result `index.html`.

## Goals of this bot

This sample starts with the [full-bundle CDN sample](./../01.a.getting-started-full-bundle/README.md) as the base template.

In this sample, we will switch the protocol in Web Chat. Direct Line Speech is a channel for communicating with a bot via a low-latency speech-focused protocol. The channel is designed for real-time speech purpose. Thus, not all features in Web Chat are supported in this channel.

> (TBD, list of limitations)

Author: Later. We have #2573.

```diff
- const res = await fetch('https://webchat-mockbot.azurewebsites.net/directline/token', { method: 'POST' });
- const { token } = await res.json();
+ const adapters = await window.WebChat.createDirectLineSpeechAdapters({
+   fetchCredentials: {
+     region: 'westus2',
+     subscriptionKey: 'YOUR_COGNITIVE_SERVICE_KEY'
+   }
+ });

  window.WebChat.renderWebChat(
    {
-     directLine: window.WebChat.createDirectLine({ token })
+     ...adapters
    },
    document.getElementById('webchat')
  );
```

> For list of supported options, you can refer to our [README on Direct Line Speech SDK](https://github.com/microsoft/BotFramework-WebChat/tree/master/packages/directlinespeech/).

Now, you should be able to communicate with your bot using the Direct Line Speech channel. If you encounter any problems while connecting to the service, you should use the in-browser developer tool and investigate.

### Using authorization token

> After you verified Web Chat is communicating with your bot using Direct Line Speech channel, we will upgrade the code to use authorization token instead.

Reviewer: ```suggestion
Once you have verified that Web Chat is communicating with your bot using the Direct Line Speech channel, the next step is to upgrade the code to use the authorization token instead.
```

The authorization token is generated using subscription key. To secure the token, the exchange must be done on server-side and the subscription key should never be passed to the browser.

> First, create an async function `fetchCredentials`. When called, it will fetch the authorization token from the server. In this sample, we will fetch the token from our demo bot. The function will return `authorizationToken` and `region` in a JavaScript object.

Reviewer: ```suggestion
First, create an async function `fetchCredentials`. When called, it will fetch the authorization token from the server. In this sample, the token is fetched from the demo bot, Mock Bot. The function will return `authorizationToken` and `region` in a JavaScript object.
```

Author: I know, we should use less "we will" in education part of sample.

```js
const fetchCredentials = async () => {
  const res = await fetch('https://webchat-mockbot-streaming.azurewebsites.net/speechservices/token', {
    method: 'POST'
  });

  if (!res.ok) {
    throw new Error('Failed to fetch authorization token and region.');
  }

  const { region, token: authorizationToken } = await res.json();

  return { authorizationToken, region };
};
```

Then, pass the function as an option to the Direct Line Speech SDK. The name of the option is `fetchCredentials`.

```diff
  const adapters = await window.WebChat.createDirectLineSpeechAdapters({
-   fetchCredentials: {
-     region: 'westus2',
-     subscriptionKey: 'YOUR_COGNITIVE_SERVICE_KEY'
-   }
+   fetchCredentials
  });

  window.WebChat.renderWebChat(
    {
      ...adapters
    },
    document.getElementById('webchat')
  );
```

> Open the web page and investigate the network traffic. You should see Web Chat is connecting to Cognitive Services using the authorization token fetched from the server.

Reviewer: ```suggestion
Open the web page and investigate the network traffic. Confirm that Web Chat is connecting to Cognitive Services using the authorization token fetched from the server.
```

## Completed code

```diff
<!DOCTYPE html>
<html lang="en-US">

<head>
  <title>Web Chat: Using Direct Line Speech</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <!--
    This CDN points to the latest official release of Web Chat. If you need to test against Web Chat's latest bits, please refer to pointing to Web Chat's MyGet feed:
    https://github.com/microsoft/BotFramework-WebChat#how-to-test-with-web-chats-latest-bits

>   -->

Reviewer: Delete comments from Readme completed code. (Keep them in the index though)
